### PR TITLE
Adds metric showing how long Query spent in Queue

### PR DESCRIPTION
**Pull request by underhood**
##### Summary
Adds metric showing how long Cloud Query spent sitting in Queue until it got picked up by Query Thread for processing.
Additionally generalizes some code to prevent code duplication.

##### Component Name
ACLK
##### Test Plan
Look at the Charts and smile

### CODE REVIEW

File: aclk/aclk_query.c [hunk 1]
```diff
@@ -319,24 +319,17 @@ static char *aclk_encode_response(char *src, size_t content_size, int keep_newli
 #pragma region ACLK_QUERY
 #endif
 
-static usec_t aclk_web_api_request_v1(RRDHOST *host, struct web_client *w, char *url)
+static usec_t aclk_web_api_request_v1(RRDHOST *host, struct web_client *w, char *url, usec_t q_created)
 {
-    usec_t t;
+    usec_t t_db = now_monotonic_high_precision_usec();
```
Comment (stelfrag) [inline aclk/aclk_query.c:324]: I think its better to move this just before the `web_client_api_request_v1` call, to measure just the request call and not any delay that can happen within the `aclk_metric_mat_update`

File: aclk/aclk_query.c [hunk 2]
```diff
@@ -427,7 +420,7 @@ static int aclk_execute_query_v2(struct aclk_query *this_query)
     mysep = strrchr(this_query->query, '/');
 
     // execute the query
-    t = aclk_web_api_request_v1(localhost, w, mysep ? mysep + 1 : "noop");
+    t = aclk_web_api_request_v1(localhost, w, mysep ? mysep + 1 : "noop", this_query->created);
```
Comment (stelfrag) [inline aclk/aclk_query.c:423]: The return value `t` is not used. Actually there is another call to `aclk_web_api_request_v1`  L357 that drops the return value (the metrics are updated  within the aclk_web_api_request_v1)  - - consider switching to void?

Comment (underhood) [reply aclk/aclk_query.c:423]: It is used here:
`buffer_sprintf(local_buffer, ",\"t-exec\": %llu,\"t-rx\": %llu,\"http-code\": %d", t, this_query->created, w->response.code);`
The aclk v2 adds fields so cloud has this information into the message.
ACLK v1 doesn't use this time

File: aclk/aclk_query.c (hunk 1 above)
Comment (underhood) [reply aclk/aclk_query.c]: resolved in latest comit

File: aclk/aclk_stats.c [hunk 3]
```diff
@@ -20,6 +20,51 @@ struct aclk_metrics aclk_metrics = {
...
+
+    .cloud_q_recvd_to_processed = { .name = "aclk_cloud_q_recvd_to_processed",
+                                    .prio = 200007,
+                                    .st = NULL,
+                                    .rd_avg = NULL,
+                                    .rd_max = NULL,
+                                    .rd_total = NULL,
+                                    .unit = "us",
+                                    .title = "Time from receiving the Cloud Query until it was picked up "
+                                             "by query thread (just before passing the to database)." }
```
Comment (stelfrag) [inline aclk/aclk_stats.c:52]: `the to database`. Needs to be rephrased.

Comment (underhood) [reply aclk/aclk_stats.c:52]: thanks fixed

**Review by stelfrag — APPROVED** [review]

**Review by stelfrag — DISMISSED** [review]

File: aclk/aclk_query.c [hunk 4]
```diff
@@ -319,24 +319,18 @@ static char *aclk_encode_response(char *src, size_t content_size, int keep_newli
...
 
-    if (aclk_stats_enabled) {
-        ACLK_STATS_LOCK;
-        aclk_metrics_per_sample.cloud_q_process_total += t;
-        aclk_metrics_per_sample.cloud_q_process_count++;
-        if (aclk_metrics_per_sample.cloud_q_process_max < t)
-            aclk_metrics_per_sample.cloud_q_process_max = t;
-        ACLK_STATS_UNLOCK;
-    }
+    aclk_metric_mat_update(&aclk_metrics_per_sample.cloud_q_db_query_time, t_db);
```
Comment (mfundul) [inline aclk/aclk_query.c:331]: You are mixing and matching realtime with monotonic time here. The first mat_update uses real time (`now_realtime_usec() - q_created`) and the second one uses monotonic time.

File: aclk/mqtt.c [hunk 5]
```diff
@@ -44,15 +44,7 @@ void publish_callback(struct mosquitto *mosq, void *obj, int rc)
...
-    if (aclk_stats_enabled) {
-        ACLK_STATS_LOCK;
-        if (aclk_metrics_per_sample.latency_max < diff)
-            aclk_metrics_per_sample.latency_max = diff;
-
-        aclk_metrics_per_sample.latency_total += diff;
-        aclk_metrics_per_sample.latency_count++;
-        ACLK_STATS_UNLOCK;
-    }
+    aclk_metric_mat_update(&aclk_metrics_per_sample.latency, diff);
```
Comment (mfundul) [inline aclk/mqtt.c:47]: Shouldn't this be inside a ` #ifdef NETDATA_INTERNAL_CHECKS`?

File: aclk/aclk_query.c (hunk 4 above)
Comment (underhood) [reply aclk/aclk_query.c:331]: the `q_created` is realtime set here `new_query->created = now_realtime_usec();` in fnc `aclk_queue_query`

Comment (underhood) [reply aclk/aclk_query.c:331]: so first `metric_mat_update` compares realtime to realtime and second `metric_mat_update` compares monotonic to monotonic...

File: aclk/mqtt.c (hunk 5 above)
Comment (underhood) [reply aclk/mqtt.c:47]: I see this:
```
void publish_callback(struct mosquitto *mosq, void *obj, int rc)
{
    UNUSED(mosq);
    UNUSED(obj);
    UNUSED(rc);
#ifdef NETDATA_INTERNAL_CHECKS
    struct timeval now, *orig;
    now_realtime_timeval(&now);
    orig = &sendTimes[ rc & 0x3ff ];
    int64_t diff = (now.tv_sec - orig->tv_sec) * USEC_PER_SEC + (now.tv_usec - orig->tv_usec);
    diff /= 1000;

    info("Publish_callback: mid=%d latency=%" PRId64 "ms", rc, diff);

    aclk_metric_mat_update(&aclk_metrics_per_sample.latency, diff);
#endif
    return;
}
```

**Review by mfundul — APPROVED** [review]

**Review by stelfrag — APPROVED** [review]

**Pull request merged**